saleor/graphql/discount/types/promotions.py

```diff
@@ -99,3 +115,90 @@ def resolve_promotion(root: models.PromotionRule, info: ResolveInfo):
     @staticmethod
     def resolve_channels(root: models.PromotionRule, info: ResolveInfo):
         return ChannelsByPromotionRuleIdLoader(info.context).load(root.id)
+
+
+class BasePromotionEvent(ModelObjectType[models.PromotionEvent]):
+    id = graphene.GlobalID()
+    date = graphene.DateTime(description="Date when event happened.")
+    type = PromotionEventsEnum(description="Promotion event type.")
+    created_by = PermissionsField(
+        UserOrApp,
+        description="User or App that created the promotion event. ",
+        permissions=[AccountPermissions.MANAGE_STAFF, AppPermission.MANAGE_APPS],
+    )
+
+    class Meta:
+        description = "History log of the promotion." + ADDED_IN_315 + PREVIEW_FEATURE
+        interfaces = [relay.Node]
+        model = models.PromotionEvent
+        doc_category = DOC_CATEGORY_DISCOUNTS
+
+    @staticmethod
+    def resolve_created_by(root: models.PromotionEvent, info):
+        requester = get_user_or_app_from_context(info.context)
+        if not requester:
+            return None
+
+        def _resolve_user(user):
+            check_is_owner_or_has_one_of_perms(
+                requester,
+                user,
+                AccountPermissions.MANAGE_STAFF,
+            )
+            return user
+
+        def _resolve_app(app):
+            check_is_owner_or_has_one_of_perms(
+                requester,
+                app,
+                AppPermission.MANAGE_APPS,
+            )
+            return app
+
+        if root.user_id:
+            return (
+                UserByUserIdLoader(info.context).load(root.user_id).then(_resolve_user)
+            )
+        if root.app_id:
+            return AppByIdLoader(info.context).load(root.app_id).then(_resolve_app)
+
+        return None
```

Comment:
Where the event creator is purged from the database, both `app` and `user` fields are set to null. As a consequence, the `created_by` resolver will return `None`. This is acceptable, but it also means we lose data regarding the event creator type (if it was an `App` or `User`). Are we comfortable with this loss, or should we contemplate introducing an additional enum field, say, `creator_type`? I faced a similar issue while working on the `App` events, hence my concern.

Reply:
No need for an extra field.